tests/deepsparse/transformers/pipelines/test_text_generation.py

```diff
@@ -12,262 +12,396 @@
+        # 2. The KV Cache is never filled up
+        # 3. KV Cache managed externally
+
+        if self.use_deepsparse_cache:
+            pytest.skip(
+                "Cannot run ORT pipeline with the internal deepsparse cache enabled."
             )
-            x = x[:, :, first_non_blank_cache_entry:, :]
-            x = x[:, :, :num_prompt_tokens, :]
-
-            """
-            y will be a cache array
-            [blank, blank, ..., prompt_cache_1, prompt_cache_2, ...]
-            we need to keep the prompt_cache entries only
-            """
-            y = y[:, :, -num_prompt_tokens:, :]
-
-            assert np.allclose(x, y, atol=1e-4)
-
-    def _get_output_huggingface(self, sequences, model_name):
-        hf_outputs = []
-        # setup tokenizer
-        tokenizer = AutoTokenizer.from_pretrained(model_name)
-        tokenizer.padding_side = "left"
-        if tokenizer.pad_token is None:
-            tokenizer.pad_token = tokenizer.eos_token
-
-        # setup model
-        model = AutoModelForCausalLM.from_pretrained(model_name)
-
-        # generate ground truth output
-        for prompt in sequences:
-            input_ids = tokenizer(prompt, return_tensors="pt").input_ids
-            generated_ids = model.generate(
-                input_ids, max_new_tokens=self.max_generated_tokens
+        model_stub, _, _, torch_ground_truth = setup
+        pipeline = Pipeline.create(
+            task="text_generation",
+            model_path=model_stub,
+            sequence_length=self.sequence_length,
+            prompt_processing_sequence_length=self.prompt_processing_sequence_length,
+            max_generated_tokens=self.num_tokens_generate,
+            force_max_tokens=True,
+            engine_type="onnxruntime",
+        )
+        output = pipeline(
+            sequences=self.prompt, return_logits=True, include_prompt_logits=True
+        )
+        cache_session = pipeline.engine.kv_cache
+        assert cache_session.total_num_processed_tokens < self.sequence_length
+        self._test_output(
+            output=output,
+            cache_session=cache_session,
+            torch_ground_truth=torch_ground_truth,
+        )
+
+    def test_ort_generation_after_kv_cache_has_been_filled(self, setup):
+        # Test the pipeline that uses ORT engine. The test covers the
+        # following scenario:
+        # 1. Prompt preprocessing is performed by multi-token engine
+        # 2. The KV Cache is filled up (old entries are removed)
+        # 3. KV Cache managed externally
+
+        if self.use_deepsparse_cache:
+            pytest.skip(
+                "Cannot run ORT pipeline with the internal deepsparse cache enabled."
             )
-            hf_output = tokenizer.decode(generated_ids[0], skip_special_tokens=True)
-            hf_outputs.append(hf_output)
-        return hf_outputs
+        model_stub, _, _, torch_ground_truth = setup
+        pipeline = Pipeline.create(
```

Comment:
Ton of `Pipeline.create` calls throughout our tests. Is there anyway we could unify some of them/store it in fixtures?

Reply:
Unifying is difficult, since each of those pipelines will differ significantly from the other.

Reply:
Did a nice trick to be less wasteful. (see get_prompt func)